skimage/color/colorconv.py

```diff
@@ -1046,15 +1046,10 @@ def gray2rgba(image, alpha=None, *, channel_axis=-1):
         RGBA image. A new dimension of length 4 is added to input
         image shape.
     """
-
     arr = np.asarray(image)
-
-    alpha_min, alpha_max = dtype_limits(arr, clip_negative=False)
-
     if alpha is None:
-        alpha = alpha_max
-
-    if not np.can_cast(alpha, arr.dtype):
```

Comment:
Highlighting this bit. As of NumPy 2.0, `np.can_cast` no longer supports Python scalars. I think [`np.result_type`](https://numpy.org/devdocs/reference/generated/numpy.result_type.html) still takes values into account when determining the resulting type because `np.min_scalar_type` is used under the hood. So we should be able to get away by comparing if the resulting type is the same as the target type `arr.dtype`. 

@seberg, taking the liberty to Cc you here. :pray: I'm sure I haven't grokked [NEP 50](https://numpy.org/neps/nep-0050-scalar-promotion.html#backward-compatibility) in its entirety yet.

Reply:
No, that isn't correct, result types of course does *not* take the value into account.

It ignores the values (same as if it was 0) for Python complex/float/ints.  For out-of-bound *integers* that would mean that the `full` call below should fail if it is out of range, though.  The `result_type` I guess still rejects floats if the array is integer.

Reply:
Oh, you seem to be right. I think I was led astray by this bit from the [docstring](https://numpy.org/devdocs/reference/generated/numpy.result_type.html#numpy.result_type)

> Otherwise, [min_scalar_type](https://numpy.org/devdocs/reference/generated/numpy.min_scalar_type.html#numpy.min_scalar_type) is called on each scalar, and the resulting data types are all combined with [promote_types](https://numpy.org/devdocs/reference/generated/numpy.promote_types.html#numpy.promote_types) to produce the return value.

Concering `np.full`, do I misunderstand you? It seems like `np.full` happily overflows out of bounds integers. E.g.
```python
np.full(1, -1, dtype=np.uint8)
# returns array([255], dtype=uint8)
np.full(1, 300, dtype=np.uint8)
# returns array([44], dtype=uint8)
```

Is there a an alternative solution in NumPy for what `np.can_cast` formerly did here? I guess I could just check if the cast is equal to the original value (https://github.com/scikit-image/scikit-image/pull/7288/commits/54a901fa1b954a9d101341117de810ad2bc583e8). I imagine the performance impact will be low as I expect alpha to be a scalar usally.

Reply:
Also curious about `np.min_scalar_type(np.iinfo(np.int64).max)` returning `dtype('uint64')`. I would have expected `np.int64` but I guess the promotion path reaches `np.uint64` first?

Reply:
arg, you are right, full is weird `np.array(value, dtype=dtype)` would work, or assignment, but full unfortunately does `np.asarray(value)` and *then* assigns.